internal/lsp/language_server/server.go

```diff
@@ -0,0 +1,71 @@
+package language_server
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"sync"
+
+	"github.com/sourcegraph/jsonrpc2"
+)
+
+type ServerArgs[State any] struct {
+	InitialState func(notify func(method string, params any)) State
+	Handler      func(r jsonrpc2.Request, state State) any
+}
+
+func sendNotification(writeMutex *sync.Mutex, method string, params any) {
+	bytes, err := json.Marshal(params)
+	if err != nil {
+		panic(err)
+	}
+	rawParams := json.RawMessage(bytes)
+	encoded := encodeMessage(jsonrpc2.Request{
+		Notif:  true,
+		Method: method,
+		Params: &rawParams,
+	})
+
+	writeMutex.Lock()
+	os.Stdout.Write([]byte(encoded))
+	writeMutex.Unlock()
+}
+
+func RunServer[State any](args ServerArgs[State]) {
+	buf := newMessageBuffer(os.Stdin)
+	mu := sync.Mutex{}
+
+	state := args.InitialState(func(method string, params any) {
```

Comment:
This file was copied-paste and moved in its own module.
The main difference is that the "sendNotification" function, instead of being called directly, is stored by the server consumer (the lsp module) in its state, so that the handlers can call that instead.
The instance of the state will be passed as value, not reference, in order to discourage writing there mutable state directly, which will be accessed concurrently.
The consumer will have to handle concurrent access manually by storing pointers explicitly